modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1702,264 +1719,549 @@ def tree_reduce(
+            }
+            new_row_lengths = lengths[0]
+            new_col_widths = lengths[1]
+        else:
+            new_row_lengths = None
+            new_col_widths = None
+        return self.__constructor__(
+            new_partitions, new_index, new_columns, new_row_lengths, new_col_widths
+        )
 
-        Notes
-        -----
-        During the join, this dataframe is considered the left, while the other is
-        treated as the right.
+    @lazy_metadata_decorator(apply_axis="both")
+    def map_select_indices_both_axes(
+        self,
+        func,
+        row_labels,
+        col_labels,
+        new_index,
+        new_columns,
+        item_to_distribute=no_default,
+    ):
+        """
+        Apply a function for a subset of the data identified by both row and column labels.
 
-        Only inner joins, left outer, right outer, and full outer joins are currently supported.
-        Support for other join types (e.g. natural join) may be implemented in the future.
+        Parameters
+        ----------
+        func : callable
+            The function to apply.
+        row_labels : list-like, default: None
+            The row labels to apply over. Must be provided with
+            `col_labels` to apply over both axes.
+        col_labels : list-like, default: None
+            The column labels to apply over. Must be provided
+            with `row_labels` to apply over both axes.
+        new_index : list-like
+            The index of the result.
+        new_columns : list-like
+            The columns of the result.
+        item_to_distribute : np.ndarray or scalar, default: no_default
+            The item to split up so it can be applied over both axes.
+
+        Returns
+        -------
+        PandasDataframe
+            A new dataframe.
         """
-        pass
+        # We are applying over both axes here, so make sure we have all the right
+        # variables set.
+        assert row_labels is not None and col_labels is not None
+        assert item_to_distribute is not no_default
+        row_partitions_list = self._get_dict_of_block_index(0, row_labels).items()
+        col_partitions_list = self._get_dict_of_block_index(1, col_labels).items()
+        new_partitions = self._partition_mgr_cls.map_select_indices_both_axes(
+            self._partitions,
+            func,
+            row_partitions_list,
+            col_partitions_list,
+            item_to_distribute,
+            # Passing caches instead of values in order to not trigger shapes recomputation
+            # if they are not used inside this function.
+            self._row_lengths_cache,
+            self._column_widths_cache,
+        )
+        return self.__constructor__(
+            new_partitions,
+            new_index,
+            new_columns,
+            self._row_lengths_cache,
+            self._column_widths_cache,
+        )
 
-    def rename(
+    def window(
```

Comment:
why we're adding this here if there's no implementation? Shouldn't it be located in the base class then?